### PR TITLE
Add support for Postgres 10

### PR DESCRIPTION
**Pull request by mtuncer**
There are some API changes regarding process utility hooks,
raw_parser(), and DoCopy() function.
Added conditional conditional compilation definitions to
make cstore_fdw run with PG 10 without breaking compatibility for
previously supported PG versions (9.3, 9.4, 9.5, 9.6)

### CODE REVIEW

File: cstore_fdw.c [hunk 1]
```diff
@@ -535,14 +608,36 @@ CopyOutCStoreTable(CopyStmt* copyStatement, const char* queryString)
...
+	/*
+	 * raw_parser returns list of RawStmt* in PG 10+ we need to
+	 * extract actual query from it.
+	 */
+	{
+		ParseState *pstate = make_parsestate(NULL);
+		RawStmt *rawStatement = (RawStmt *) rawQuery;
+
+		pstate->p_sourcetext = newQuerySubstring->data;
+
```
Comment (lithp) [inline cstore_fdw.c]: Probably don't need this many newlines, this one seems superfluous

File: cstore_fdw.c [hunk 2]
```diff
@@ -55,15 +55,41 @@
 #include "utils/tqual.h"
 
 
+#if PG_VERSION_NUM < 100000
+/*
+ * QueryEnvironment definition only exist in PG_VERSION >= 10
+ * Here is a dummy definition to enable building in previous versions.*/
+typedef void QueryEnvironment;
```
Comment (lithp) [inline cstore_fdw.c]: You no longer need this now that you've moved every instance of QueryEnvironment into an `#if PG_VERSION_NUM`

**Comment by lithp** [comment]
Wow, thanks for the update! I expected you to push back a little :)

I still haven't tested this yet, sorry. I'm about to hop on a plane but I'll do a sanity check first thing Tuesday and probably approve the changes then. Does that block you at all?

**Comment by mtuncer** [comment]
@lithp its fine. I will push another commit before you look at it.

**Comment by lithp** [comment]
Apologies for taking a while to look at this!

When I build on pg10 I get a segabort while running `COPY TO` on the table.

The commands I ran:

```
CREATE EXTENSION cstore_fdw;
CREATE SERVER cstore_server FOREIGN DATA WRAPPER cstore_fdw;
CREATE FOREIGN TABLE customer_reviews (
    customer_id TEXT,
    review_date DATE,
    review_rating INT,
    review_votes INT,
    review_helpful_votes INT,
    product_id CHAR(10),
    product_title TEXT,
    product_sales_rank BIGINT,
    product_group TEXT,
    product_category TEXT,
    product_subcategory TEXT,
    similar_product_ids CHAR(10)[]
) SERVER cstore_server OPTIONS(compression 'pglz');
COPY customer_reviews FROM '/home/brian/Work/pg-10-debug/customer_reviews_1998.csv' WITH CSV;
COPY customer_reviews FROM '/home/brian/Work/pg-10-debug/customer_reviews_1999.csv' WITH CSV;
COPY customer_reviews TO '/home/brian/Work/pg-10-debug/customer_reviews_all.csv' WITH CSV;
```

GDB spits out this:

```
#2  0x0000000000871cdb in ExceptionalCondition (conditionName=conditionName@entry=0x7f85857d2200 "!(((((const Node*)(minNode))->type) == T_Const))", 
    errorType=errorType@entry=0x7f85857d15fc "FailedAssertion", fileName=fileName@entry=0x7f85857d2081 "cstore_reader.c", lineNumber=lineNumber@entry=942) at assert.c:54
#3  0x00007f85857ce759 in UpdateConstraint (baseConstraint=baseConstraint@entry=0x10e7450, minValue=17662432, maxValue=17662488) at cstore_reader.c:942
#4  0x00007f85857cf3c4 in SelectedBlockMask (stripeSkipList=stripeSkipList@entry=0x10e4520, projectedColumnList=projectedColumnList@entry=0xfc6270, 
    whereClauseList=whereClauseList@entry=0x0) at cstore_reader.c:764
#5  0x00007f85857cf483 in LoadFilteredStripeBuffers (tableFile=0xf02500, stripeMetadata=0xfcbd48, tupleDescriptor=<optimized out>, projectedColumnList=0xfc6270, 
    whereClauseList=0x0) at cstore_reader.c:485
#6  0x00007f85857cf5d3 in CStoreReadNextRow (readState=readState@entry=0xfcc4b8, columnValues=columnValues@entry=0xfc8fd8, 
    columnNulls=columnNulls@entry=0xfc9070 '\001' <repeats 12 times>, "~") at cstore_reader.c:260
#7  0x00007f85857cb506 in CStoreIterateForeignScan (scanState=<optimized out>) at cstore_fdw.c:1816
#8  0x000000000064bb68 in ForeignNext (node=node@entry=0xfc89f8) at nodeForeignscan.c:54
#9  0x000000000062b850 in ExecScanFetch (recheckMtd=0x64ba29 <ForeignRecheck>, accessMtd=0x64bb1d <ForeignNext>, node=0xfc89f8) at execScan.c:95
#10 ExecScan (node=node@entry=0xfc89f8, accessMtd=accessMtd@entry=0x64bb1d <ForeignNext>, recheckMtd=recheckMtd@entry=0x64ba29 <ForeignRecheck>) at execScan.c:143
#11 0x000000000064bbba in ExecForeignScan (node=node@entry=0xfc89f8) at nodeForeignscan.c:119
#12 0x000000000062a4bb in ExecProcNode (node=node@entry=0xfc89f8) at execProcnode.c:501
#13 0x0000000000625526 in ExecutePlan (estate=estate@entry=0xfc87e0, planstate=0xfc89f8, use_parallel_mode=0 '\000', operation=operation@entry=CMD_SELECT, 
    sendTuples=sendTuples@entry=1 '\001', numberTuples=numberTuples@entry=0, direction=ForwardScanDirection, dest=0xfc6638, execute_once=1 '\001') at execMain.c:1696
#14 0x0000000000626386 in standard_ExecutorRun (queryDesc=0xfc6690, direction=ForwardScanDirection, count=0, execute_once=execute_once@entry=1 '\001') at execMain.c:360
#15 0x000000000062640f in ExecutorRun (queryDesc=<optimized out>, direction=direction@entry=ForwardScanDirection, count=count@entry=0, 
    execute_once=execute_once@entry=1 '\001') at execMain.c:303
#16 0x00000000005acf99 in CopyTo (cstate=cstate@entry=0xf821a0) at copy.c:2072
#17 0x00000000005ad120 in DoCopyTo (cstate=cstate@entry=0xf821a0) at copy.c:1897
#18 0x00000000005b368e in DoCopy (pstate=pstate@entry=0xf051a8, stmt=stmt@entry=0xf5f030, stmt_location=stmt_location@entry=-1, stmt_len=stmt_len@entry=-1, 
    processed=processed@entry=0x7ffe1f0e0720) at copy.c:983
#19 0x00007f85857ca99a in CopyOutCStoreTable (copyStatement=0xf5f030, queryString=<optimized out>) at cstore_fdw.c:584
#20 0x00007f85857ca9e8 in CStoreProcessCopyCommand (copyStatement=copyStatement@entry=0xf5f030, 
    queryString=queryString@entry=0xf5e410 "COPY customer_reviews TO '/home/brian/Work/pg-10-debug/customer_reviews_all.csv' WITH CSV;", 
```

**Comment by lithp** [comment]
Murat and I tested this together and got my machine into a working state. Our best guess is that I didn't run `make clean` before building so it's possible some of the files were still built against 9.6. Anyway, it works now, I say ship it!

**Pull request merged**